File: app/src/main/java/de/project/lukas/model/DevicesManager.java
package de.project.lukas.model;

import android.annotation.SuppressLint;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothManager;
import android.bluetooth.le.BluetoothLeScanner;
import android.bluetooth.le.ScanCallback;
import android.bluetooth.le.ScanResult;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Handler;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;

@SuppressLint("MissingPermission")
public final class DevicesManager extends ViewModel {
    private static final DevicesManager instance = new DevicesManager();
    private final MutableLiveData<List<Device>> devices = new MutableLiveData<>(new ArrayList<>());   // verbundene Devices
    private final MutableLiveData<Boolean> isScanning = new MutableLiveData<>(false);
    private BluetoothAdapter bluetoothAdapter;
    private BluetoothManager bluetoothManager;
    private BluetoothLeScanner bluetoothScanner;

    private final ScanCallback scanCallback = new ScanCallback() {
        @Override
        public void onScanResult(int callbackType, ScanResult result) {
            BluetoothDevice device = result.getDevice();

            if (device.getName() == null) {
                return;
            }

            if (!hasDevice(device.getAddress())) {
                if (TrainHub.canConnect(result)) {
                    addDevice(new TrainHub(device));
                } else if (TrainBase.canConnect(result)) {
                    addDevice(new TrainBase(device));
                } else if (Remote.canConnect(result)) {
                    addDevice(new Remote(device));
                } else if (Switch.canConnect(result)) {
                    addDevice((new Switch(device)));
                }
            }
        }
    };

    public static DevicesManager getInstance() {
        return instance;
    }

    public LiveData<List<Device>> getDevices() {
        return devices;
    }

    public LiveData<Boolean> isScanning() {
        return isScanning;
    }

    public boolean isTesting() {   return false;    }

    public boolean isBluetoothEnabled() {
        return bluetoothAdapter != null && bluetoothAdapter.isEnabled();
    }

    @SuppressLint("DefaultLocale")
    private DevicesManager() {
        if (isTesting()) {
            for (int i = 0; i < 6; i++) {
                addDevice(new Switch(String.format("Switch #%d", i)));
                addDevice(new Remote(String.format("Remote #%d", i)));
                addDevice(new TrainHub(String.format("Train Hub #%d", i)));
                addDevice(new TrainBase(String.format("Train Base #%d", i)));
            }
        }
    }

    public void setBluetoothManager(BluetoothManager bluetoothManager) {
        this.bluetoothManager = bluetoothManager;
    }

    public void startScanning() {
        if (Boolean.TRUE.equals(isScanning.getValue())) {
            return;
        }

        bluetoothAdapter = bluetoothManager.getAdapter();

        if (bluetoothAdapter == null || !bluetoothAdapter.isEnabled()) {
            return;
        }

        bluetoothScanner = bluetoothAdapter.getBluetoothLeScanner();

        if (bluetoothScanner == null) {
            return;
        }

        isScanning.postValue(true);

        AsyncTask.execute(() -> bluetoothScanner.startScan(scanCallback));

        // Stop the scanning automatically after 20 sec.
        new Handler().postDelayed(this::stopScanning, 20000);
    }

    public void stopScanning() {
        if (Boolean.FALSE.equals(isScanning.getValue())) {
            return;
        }

        isScanning.postValue(false);

        AsyncTask.execute(() -> bluetoothScanner.stopScan(scanCallback));
    }

    private boolean hasDevice(String address) {
        List<Device> deviceList = Objects.requireNonNull(devices.getValue());

        for (Device device : deviceList) {
            if (device.getAddress().equals(address)) {
                return true;
            }
        }

        return false;
    }

    public void addDevice(Device deviceToAdd) {
        List<Device> deviceList = Objects.requireNonNull(devices.getValue());

        deviceList.add(deviceToAdd);
        deviceToAdd.getName().observeForever(v -> sortDevices());

        deviceToAdd.getStatus().observeForever(isConnected -> {
            if (!isConnected) {
                removeDevice(deviceToAdd);
            }
        });

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            deviceList.sort(Comparator.comparing((Device t) -> t.getClass().getName(), Comparator.reverseOrder()).thenComparing(Device::toString));
        }

        devices.postValue(deviceList);
    }

    public void removeDevice(Device device) {
        device.disconnect();

        List<Device> deviceList = Objects.requireNonNull(devices.getValue());

        if (!deviceList.remove(device)) {
            return;
        }

        devices.postValue(deviceList);
    }

    public void switchOffDevice(Device device) {
        device.switchOff();

        List<Device> deviceList = Objects.requireNonNull(devices.getValue());

        if (!deviceList.remove(device)) {
            return;
        }

        devices.postValue(deviceList);
    }

    private void sortDevices() {
        List<Device> deviceList = devices.getValue();

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && deviceList != null) {
            deviceList.sort(Comparator.comparing((Device t) -> t.getClass().getName(), Comparator.reverseOrder()).thenComparing(Device::toString));
        }

        devices.postValue(deviceList);
    }

    public void AllTrainMotorStop() {
        List<Device> deviceList = Objects.requireNonNull(devices.getValue());

        for (Device device : deviceList) {
            if (device instanceof TrainHub) {
                ((TrainHub) device).motorStop();
            }
            if (device instanceof TrainBase) {
                ((TrainBase) device).motorStop();
            }
        }
    }

    @SuppressWarnings("unused")
    public void disconnectAll() {
        List<Device> deviceList = Objects.requireNonNull(devices.getValue());

        for (Device device : deviceList) {
            device.disconnect();
        }
    }
}

File: app/src/main/java/de/project/lukas/model/DevicesManager.kt
package de.project.lukas.model

import android.annotation.SuppressLint
import android.app.Application
import android.bluetooth.BluetoothManager
import android.bluetooth.le.BluetoothLeScanner
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * Owns the BLE scan and the set of connected devices. As an [AndroidViewModel] it survives
 * configuration changes but is bound to the Activity's lifecycle (no static singleton), and it
 * cleans up its per-device collectors instead of leaking [kotlinx.coroutines.flow.StateFlow]
 * observers.
 */
@SuppressLint("MissingPermission")
class DevicesManager(application: Application) : AndroidViewModel(application) {

    private val bluetoothAdapter =
        application.getSystemService(BluetoothManager::class.java)?.adapter
    private var scanner: BluetoothLeScanner? = null

    private val _devices = MutableStateFlow<List<Device>>(emptyList())
    val devices: StateFlow<List<Device>> = _devices.asStateFlow()

    private val _isScanning = MutableStateFlow(false)
    val isScanning: StateFlow<Boolean> = _isScanning.asStateFlow()

    private val deviceJobs = mutableMapOf<Device, Job>()

    val isBluetoothEnabled: Boolean
        get() = bluetoothAdapter?.isEnabled == true

    private val scanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            val bluetoothDevice = result.device
            bluetoothDevice.name ?: return
            if (hasDevice(bluetoothDevice.address)) return

            val context = getApplication<Application>()
            when {
                TrainHub.canConnect(result) -> addDevice(TrainHub(context, bluetoothDevice))
                TrainBase.canConnect(result) -> addDevice(TrainBase(context, bluetoothDevice))
                Remote.canConnect(result) -> addDevice(Remote(context, bluetoothDevice))
                Switch.canConnect(result) -> addDevice(Switch(context, bluetoothDevice))
            }
        }
    }

    fun startScanning() {
        if (_isScanning.value) return
        val adapter = bluetoothAdapter ?: return
        if (!adapter.isEnabled) return
        val leScanner = adapter.bluetoothLeScanner ?: return

        scanner = leScanner
        _isScanning.value = true
        leScanner.startScan(scanCallback)

        // Stop scanning automatically after 20 seconds.
        viewModelScope.launch {
            delay(SCAN_DURATION_MS)
            stopScanning()
        }
    }

    fun stopScanning() {
        if (!_isScanning.value) return
        _isScanning.value = false
        scanner?.stopScan(scanCallback)
    }

    fun removeDevice(device: Device) {
        device.disconnect()
        detach(device)
    }

    fun switchOffDevice(device: Device) {
        device.switchOff()
        detach(device)
    }

    fun stopAllTrains() {
        _devices.value.forEach { device ->
            when (device) {
                is TrainHub -> device.motorStop()
                is TrainBase -> device.motorStop()
            }
        }
    }

    private fun addDevice(device: Device) {
        _devices.update { (it + device).sortedForDisplay() }

        deviceJobs[device] = viewModelScope.launch {
            // Re-sort when a device is renamed.
            launch { device.name.collect { _devices.update { list -> list.sortedForDisplay() } } }
            // Drop the device automatically once it disconnects.
            device.isConnected.collect { connected ->
                if (!connected) detach(device)
            }
        }
    }

    private fun detach(device: Device) {
        deviceJobs.remove(device)?.cancel()
        _devices.update { it - device }
    }

    private fun hasDevice(address: String) = _devices.value.any { it.address == address }

    private fun List<Device>.sortedForDisplay(): List<Device> = sortedWith(
        compareByDescending<Device> { it.javaClass.name }.thenBy { it.name.value },
    )

    override fun onCleared() {
        super.onCleared()
        _devices.value.forEach { it.disconnect() }
    }

    companion object {
        private const val SCAN_DURATION_MS = 20_000L
    }
}
